### ISSUE TITLE
Use json streaming API for reading in json data blobs.

### ISSUE THREAD

**Issue by kevingranade** [CleverRaven/Cataclysm-DDA#12744]
Reading in static json data dominates our startup time, something like 10 seconds of it on my system.
We could possibly cut that nearly in half by using the streaming API.
How it works now:
In DynamicDataLoader::load_all_from_json():
-  For each json object, read until you reach the closing '}', building a map of indices from member name to the index where it occurs.
-  Read the "type" member of the object, and pass the json object to the appropriate decoder based on its value.

In each type-specific decoder:
-  Members are accessed by doing json.read("membername", place_to_write);

Options:
Have the DynamicDataLoader read the type field using the streaming API, then rewind the stream to before the object start before passing it to the specific decoder.
PROS: Doesn't require changing the JSON unless there are objects where type doesn't come first.
CONS: the rewind thing is a little weird, and we're re-doing (a small amount of) work, requires every json object to have its first field by "type", which isn't technically correct json.

Move the type field outside the json object proper, wrapping the type/object pair in an array.
Before: `{"type":"item", <contents...>}`
After: `["item", {<contents...>}]`
PROS: More 'correct' processing.
CONS: Requires changing every json entity, lots of extraneous delimiters (similar numbers of characters though).

With either route, the first step should be to pick one of the options (or another better one) and just push the creation of json object wrappers down into each type-specific decoder.
Once the initial work for this is done, loaders that are taking a long time can be profiled and switched to also use the streaming API, which is when the payoff will happen.
##

**Comment by BevapDin** [comment]
Most loading code for the game data (not saves) uses `get_*` functions which throw properly when the member does not exits or is of the wrong type. I consider this quite useful. It also has a nice syntax to specify default values.

The streaming API has one huge disadvantages: it's far more complicated to ensure a member has actually been defined in JSON.
```C++
// Non-streaming:
data1 = jobj.get_string("data1");
data2 = jobj.get_int("data2");
// data1 and data2 have definitely been loaded, otherwise json classes would have thrown

// Streaming:
while(jo.has_more()) {
    const std::string member_name = jsin.get_member_name();
    if(member_name == "data1") {
        data1 = jsin.get_string();
    } else if(member_name == "data2") {
        data2 = jsin.get_int();
    } else {
        ...
    }
}
// At this point, it's not sure whether data1 and data2 have actually been loaded,
// what if they did not exists in the JSON?
```
This requires separate bookkeeping to register the member as loaded and additional checking after the loop. And I consider not checking whether a *required* value has actually been defined a serious bug or at least very user unfriendly. This is made even worse if the member is a primitive type that was not initialized (because it would be overridden by the *mandatory* value from JSON).

----

> For each json object, read until you reach the closing '}', building a map of indices from member name to the index where it occurs.

That's what the constructor of `JsonObject` does, which is invoked every time one calls `get_object` (which creates a *new* JsonObject). This parses the whole object, recursively including all its members. Same for arrays. This means the innermost part of the JSON file are parsed (mostly just skipped via `JsonIn::skip_value`) many times.

```JSON
[
    { "type": "...", "data": { } },
    { "type": "...", "data": { } }
]
```
This JSON file, loaded with:
```C++
JsonIn jsin( stream );
JsonArray array = jsin.get_array(); // parses the whole file (1.)
while( array.has_more() ) {
    JsonObject obj = array.next_object(); // parses the object (2.)
    JsonObject data = obj.get_object("data"); // parses the data object (3.)
    
}
```
The content of the "data" member is parsed 3 times.

I think having the Json classes parse the file once (completely) and simply storing the parsed data in some way (may not be full tree of objects, but at least a tree of stream positions) would reduce that parsing overhead considerably.

**Comment by kevingranade** [comment]
I agree one extreme or the other is the way to go, if we're going to parse
a whole file of objects it should probably do a single tokenization pass,
at which point we have a tree of all of the objects and arrays (actually
pointers to them, we should never copy file data before further parsing),
and the other alternative is to go all the way streaming, which avoids
allocating all those indices and the lookups into the trees, and gets us
all the way down to one pass over the data.

**Comment by i2amroy** [comment]
If we want to check for correctness in JSON->JSON references aren't we going to need at least two stream passes regardless?

**Comment by kevingranade** [comment]
Consistency checks use the decoded objects, not json streams.

**Comment by i2amroy** [comment]
Ah, gotcha.

**Comment by illi-kun** [comment]
### This issue was closed as it appears inactive.

Reducing open issues to those which are (or will) be actively worked upon helps us focus our efforts. This issue has not been deleted - it still appears in searches and if it contains relevant information you are encouraged to continue to link to it.

### If this issue was a bug
It should be reopened if it can be reproduced in the current build. You can obtain the most recent copy [here](http://dev.narc.ro/cataclysm/jenkins-latest/). Please check there is not a more recent report of this bug before doing so. If no more recent report exists you should continue the discussion in this issue.

### If this was a feature request
If the consensus was that the idea was good you could consider submitting an implementation via a PR. If you want to comment further please do so here as opposed to opening a new issue. Before posting check nobody has already made the same point and consider whether your comments are likely to lead to an implementation. If you have doubts about either consider instead [voting](http://iv.cataclysmdda.com:8111/all) for the issue

### If you want to work on this issue
Then either assign it to yourself or if you are unable to do so claim it via adding a comment. Please don't assign others or make a general request for action.
